### PR TITLE
Port list-fonts

### LINKED ISSUE #1143: Port list-fonts
```c
DEFUN ("list-fonts", Flist_fonts, Slist_fonts, 1, 4, 0,
       doc: /* List available fonts matching FONT-SPEC on the current frame.
Optional 2nd argument FRAME specifies the target frame.
Optional 3rd argument NUM, if non-nil, limits the number of returned fonts.
Optional 4th argument PREFER, if non-nil, is a font-spec to
control the order of the returned list.  Fonts are sorted by
how close they are to PREFER.  */)
  (Lisp_Object font_spec, Lisp_Object frame, Lisp_Object num, Lisp_Object prefer)
{
  struct frame *f = decode_live_frame (frame);
  Lisp_Object vec, list;
  EMACS_INT n = 0;

  CHECK_FONT_SPEC (font_spec);
  if (! NILP (num))
    {
      CHECK_NUMBER (num);
      n = XINT (num);
      if (n <= 0)
	return Qnil;
    }
  if (! NILP (prefer))
    CHECK_FONT_SPEC (prefer);

  list = font_list_entities (f, font_spec);
  if (NILP (list))
    return Qnil;
  if (NILP (XCDR (list))
      && ASIZE (XCAR (list)) == 1)
    return list1 (AREF (XCAR (list), 0));

  if (! NILP (prefer))
    vec = font_sort_entities (list, prefer, f, 0);
  else
    vec = font_vconcat_entity_vectors (list);
  if (n == 0 || n >= ASIZE (vec))
    list = CALLN (Fappend, vec, Qnil);
  else
    {
      for (list = Qnil, n--; n >= 0; n--)
	list = Fcons (AREF (vec, n), list);
    }
  return list;
}
```

### PR DESCRIPTION
**Pull request by agraven**
Closes #1143.

### CODE REVIEW

File: rust_src/src/fonts.rs [hunk 1]
```diff
@@ -167,6 +170,48 @@ impl From<LispObject> for Option<LispFontObjectRef> {
     }
 }
 
+pub type LispFontSpecRef = ExternalPtr<Lisp_Font_Spec>;
```
Comment (db48x) [inline rust_src/src/fonts.rs]: This could just be called `FontSpecRef`, since there's no potential for confusion.

File: rust_src/src/fonts.rs [hunk 2]
```diff
@@ -265,4 +310,51 @@ pub fn font_at_lisp(
...
+/// Optional 2nd argument FRAME specifies the target frame.
+/// Optional 3rd argument NUM, if non-nil, limits the number of returned fonts.
+/// Optional 4th argument PREFER, if non-nil, is a font-spec to
+/// control the order of the returned list.  Fonts are sorted by
+/// how close they are to PREFER.
+#[lisp_fn(min = "1")]
+pub fn list_fonts(
+    font_spec: LispFontSpecRef,
+    frame: LispFrameLiveOrSelected,
+    num: Option<LispNumber>,
```
Comment (brotzeit) [inline rust_src/src/fonts.rs]: I think this should be `Option<LispNumber>`. The c code uses `CHECK_NUMBER`.

File: rust_src/src/fonts.rs [hunk 3]
```diff
@@ -265,4 +310,51 @@ pub fn font_at_lisp(
...
+
+    let n = match num.map(|n| n.to_fixnum()) {
+        Some(n) if n < 0 => return Qnil,
+        Some(n) => n,
+        None => 0,
+    } as usize;
+
+    let list = unsafe { font_list_entities(frame.as_mut(), font_spec.into()) };
+    match list.as_cons() {
+        Some(list) => {
```
Comment (nickdrozd) [inline rust_src/src/fonts.rs]: Pattern match `list` as a tuple

File: rust_src/src/fonts.rs [hunk 4]
```diff
@@ -265,4 +310,51 @@ pub fn font_at_lisp(
...
+            if list.cdr().is_nil() && list.car().force_vector().len() == 1 {
```
Comment (nickdrozd) [inline rust_src/src/fonts.rs]: `force_vector` gets called twice here. Is there a way of pattern matching it?

File: rust_src/src/fonts.rs [hunk 5]
```diff
@@ -265,4 +310,51 @@ pub fn font_at_lisp(
...
+#[lisp_fn(min = "1")]
+pub fn list_fonts(
+    font_spec: LispFontSpecRef,
+    frame: LispFrameLiveOrSelected,
+    num: Option<LispNumber>,
+    prefer: Option<LispFontSpecRef>,
+) -> LispObject {
+    let mut frame: LispFrameRef = frame.into();
+
+    let n = match num.map(|n| n.to_fixnum()) {
```
Comment (nickdrozd) [inline rust_src/src/fonts.rs]: Is there a `to_maybe_fixnum` conversion or something like that?

Comment (agraven) [reply rust_src/src/fonts.rs]: Assuming brotzeit is right and `num` is supposed to be an EmacsInt, the map won't be necessary

File: rust_src/src/fonts.rs (hunk 4 above)
Comment (agraven) [reply rust_src/src/fonts.rs]: It should be possible with an `if let`

File: rust_src/src/fonts.rs (hunk 5 above)
Comment (shaleh) [reply rust_src/src/fonts.rs]: in lisp.h `#define lisp_h_CHECK_NUMBER(x) CHECK_TYPE (INTEGERP (x), Qintegerp, x)` yep, that is a EmacsInt.

File: rust_src/src/fonts.rs (hunk 2 above)
Comment (agraven) [reply rust_src/src/fonts.rs]: Fixed

File: rust_src/src/fonts.rs (hunk 4 above)
Comment (agraven) [reply rust_src/src/fonts.rs]: Turns out you can do it even more neatly with a guarded match.

File: rust_src/src/fonts.rs (hunk 3 above)
Comment (agraven) [reply rust_src/src/fonts.rs]: Fixed

**Comment by shaleh** [comment]
Look at the Clippy messages:
```
warning: identical conversion
   --> src/fonts.rs:351:9
    |
351 | /         (0..=n)
352 | |             .into_iter()
    | |________________________^ help: consider removing `.into_iter()`: `(0..=n)`
    |
    = note: #[warn(clippy::identity_conversion)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#identity_conversion
```

File: rust_src/src/fonts.rs [hunk 6]
```diff
@@ -167,6 +169,48 @@ impl From<LispObject> for Option<LispFontObjectRef> {
...
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn font_vconcat_entity_vectors(list: LispObject) -> LispObject {
+    vconcat_entity_vectors(list.into())
+}
+
+/// Concatenate all elements of LIST into one vector. LIST is a list of font-entity vectors.
+fn vconcat_entity_vectors(list: LispCons) -> LispObject {
```
Comment (shaleh) [inline rust_src/src/fonts.rs]: My only concern with using Rust's vectors here is SAFE_ALLOCA can trigger garbage collection at the end of the function. I do not believe this code will do that as written. It is _probably_ ok.

File: rust_src/src/fonts.rs [hunk 7]
```diff
@@ -265,4 +309,50 @@ pub fn font_at_lisp(
...
+    let mut frame: LispFrameRef = frame.into();
+
+    let n = match num {
+        Some(n) if n < 0 => return Qnil,
+        Some(n) => n,
+        None => 0,
+    } as usize;
+
+    let list = unsafe { font_list_entities(frame.as_mut(), font_spec.into()) };
+    match list.into() {
```
Comment (shaleh) [inline rust_src/src/fonts.rs]: Yuck, rustfmt did not make this readable.

File: src/font.c [hunk 8]
```diff
@@ -2168,26 +2168,6 @@ font_score (Lisp_Object entity, Lisp_Object *spec_prop)
...
-  EMACS_INT nargs = XFASTINT (Flength (list));
-  Lisp_Object *args;
-  USE_SAFE_ALLOCA;
-  SAFE_ALLOCA_LISP (args, nargs);
-  ptrdiff_t i;
-
-  for (i = 0; i < nargs; i++, list = XCDR (list))
-    args[i] = XCAR (list);
-  Lisp_Object result = Fvconcat (nargs, args);
-  SAFE_FREE ();
```
Comment (shaleh) [inline src/font.c]: This line here is where GC could be triggered. `SAFE_FREE` is a light wrapper around `unbind_to`.

File: rust_src/src/fonts.rs [hunk 9]
```diff
@@ -265,4 +309,50 @@ pub fn font_at_lisp(
...
+    }
+
+    let vec = match prefer {
+        None => unsafe { font_sort_entities(list, prefer.into(), frame.as_mut(), 0) },
+        Some(_) => vconcat_entity_vectors(list.into()),
+    }
+    .force_vector();
+
+    if n == 0 || n >= vec.len() {
+        fns::append(&mut [list, vec.into(), Qnil])
```
Comment (shaleh) [inline rust_src/src/fonts.rs]: Is this doing the same as `list = CALLN (Fappend, vec, Qnil);`? Why is list being included?

Comment (agraven) [reply rust_src/src/fonts.rs]: That was a mixup on my part. List should not be included.

File: src/font.c (hunk 8 above)
Comment (agraven) [reply src/font.c]: I think the only reason the `unbind_to` call is necessary is that the `SAFE_ALLOCA_LISP` invocation adds to the unwind protect stack. It should be fine without it. It's not a function that's invoked often.

File: rust_src/src/fonts.rs (hunk 1 above)
Comment (agraven) [reply rust_src/src/fonts.rs]: I prefer to keep it as it as, as it keeps the naming consistent. I think the potential confusion over different naming styles outweighs the convenience of the slightly shorter name.

**Comment by agraven** [comment]
This should be ready for merge assuming the CI passes.

**Pull request merged**